### PR TITLE
Initial implementation of FieldBuilder

### LINKED ISSUE #11166: Search:  Design and implement FieldBuilder
In the design, consider:

* Return a mutable list of fields (#11584)
* Handle self-referential types (avoid cycles) (#8738)

### PR DESCRIPTION
**Pull request by heaths**
Fixes #11166. This is a port of the original track 1 FieldBuilder, which was merely a sample for track 2 GA.

See https://github.com/heaths/azure-sdk-for-net/pull/3 for original discussion.

### CODE REVIEW

**Comment by heaths** [comment]
@weshaggard how are we supposed to add new fully-qualified links from master when we're just adding them in a PR? Link verification failed:

image

This file will exist at that location once checked in. Seems link verification should somehow take that into account. For now I'll work around it by using a relative URL, but will have to open a new PR to then fix it once this is merged, which is a bit of a hassle.

File: sdk/search/Azure.Search.Documents/README.md
Comment (pakrym) [inline sdk/search/Azure.Search.Documents/README.md:220]: Do we need the flag at this point?

File: sdk/search/Azure.Search.Documents/Directory.Build.props
Comment (tg-msft) [inline sdk/search/Azure.Search.Documents/Directory.Build.props:43]: We can't ship with a dependency on Microsoft.Spatial.  It should have its own feature flag so we don't need to `'$(UseAzureCoreExperimentalSpatial)' != 'true'`.

File: sdk/search/Azure.Search.Documents/tests/Azure.Search.Documents.Tests.csproj
Comment (tg-msft) [inline sdk/search/Azure.Search.Documents/tests/Azure.Search.Documents.Tests.csproj:48]: When it's true we want to remove them?

File: sdk/search/Azure.Search.Documents/README.md
Comment (heaths) [reply sdk/search/Azure.Search.Documents/README.md:220]: @tg-msft and @AlexGhiondea wanted it. We'll almost definitely end up shipping it and should remove it later.

File: sdk/search/Azure.Search.Documents/samples/Sample04_FieldBuilder.md
Comment (tg-msft) [inline sdk/search/Azure.Search.Documents/samples/Sample04_FieldBuilder.md:8]: Feels kind of weird to start the sample with workarounds.  Might be better to show normal use and include this at the end?  Or at least add an intro paragraph explaining how you're decorating the properties with the attributes and what they effectively mean before we explain the enum weirdness?

File: sdk/search/Azure.Search.Documents/src/Indexes/FieldBuilder.cs
Comment (tg-msft) [inline sdk/search/Azure.Search.Documents/src/Indexes/FieldBuilder.cs:105]: Would it simplify things to remove the extra block scoped method like?

```C#
            if (ObjectInfo.TryGet(modelType, nameProvider, out ObjectInfo info) &&
                info.Properties.Length > 0)
            {
                // Use Stack to avoid a dependency on ImmutableStack for now.
                return Build(modelType, info, nameProvider, new Stack<Type>(new[] { modelType }));
            }

             string errorMessage =
                    $"Type '{modelType}' does not have properties which map to fields of an Azure Search index. Please use a " 
                    "class or struct with public properties.";
                throw new ArgumentException(errorMessage, nameof(modelType));
```

File: sdk/search/Azure.Search.Documents/src/Indexes/FieldBuilder.cs
Comment (tg-msft) [inline sdk/search/Azure.Search.Documents/src/Indexes/FieldBuilder.cs:218]: Is this the core of the changes?  Was there anything else that needs extra scrutiny?

**Review by tg-msft — APPROVED** [review]
The only ship blocker is removing Microsoft.Spatial.

File: sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs
Comment (heaths) [reply sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs:1293]: That's one reason, yes. Also so people can still easily use the object initializer pattern. The implementation will make sure it's not null. @KrzysztofCwalina was okay with this when I discussed it with him.

File: sdk/search/Azure.Search.Documents/tests/Azure.Search.Documents.Tests.csproj
Comment (heaths) [reply sdk/search/Azure.Search.Documents/tests/Azure.Search.Documents.Tests.csproj:48]: Given all .cs is included by default, it's far easier to remove what you don't want.

File: sdk/search/Azure.Search.Documents/src/Indexes/FieldBuilder.cs
Comment (heaths) [reply sdk/search/Azure.Search.Documents/src/Indexes/FieldBuilder.cs:105]: This was ported track 1 code with minimal changes desired.

File: sdk/search/Azure.Search.Documents/Directory.Build.props
Comment (heaths) [reply sdk/search/Azure.Search.Documents/Directory.Build.props:15]: #14096

File: sdk/search/Azure.Search.Documents/src/Indexes/FieldBuilder.cs
Comment (heaths) [reply sdk/search/Azure.Search.Documents/src/Indexes/FieldBuilder.cs:29]: @jongio established this previously.

File: sdk/search/Azure.Search.Documents/src/Indexes/FieldBuilder.cs
Comment (heaths) [reply sdk/search/Azure.Search.Documents/src/Indexes/FieldBuilder.cs:218]: It's the last 6 commits since you reviewed when I had it staged in my fork entirely. As of this moment:

https://github.com/Azure/azure-sdk-for-net/pull/14023/files/4b4e2bf897bb01b9b56c1aaf3d63381c34a94e53..021993263e1bfc9874e333643fa74070494c81a8

File: sdk/search/Azure.Search.Documents/Directory.Build.props
Comment (heaths) [reply sdk/search/Azure.Search.Documents/Directory.Build.props:43]: I will have to cut Microsoft.Spatial entirely, then. You had me add this to the test project when `FieldBuilder` was a sample to make sure it still works; however, with moving `FieldBuilder` to the shipping project, it came along. I agree we shouldn't ship it as a dependency, but see no other choice than to remove spatial support.

File: sdk/search/Azure.Search.Documents/samples/Sample04_FieldBuilder.md
Comment (heaths) [reply sdk/search/Azure.Search.Documents/samples/Sample04_FieldBuilder.md:8]: Fair point. I put this in samples but originally it was meant to be just a workaround, but given it's location I agree at least rewording this would be good. The sample code itself should already serve the "just the basics" well enough.

File: sdk/search/Azure.Search.Documents/Directory.Build.props
Comment (tg-msft) [reply sdk/search/Azure.Search.Documents/Directory.Build.props:43]: We're removing Spatial support elsewhere too for the same reason (i.e., SearchDocument, SearchFilter, etc.) until we can take a dependency on whatever ships in Azure.Core.  It's fine to leave with a new flag if you want it for testing, etc., but it should be off for the next release.

File: sdk/search/Azure.Search.Documents/tests/Azure.Search.Documents.Tests.csproj
Comment (tg-msft) [reply sdk/search/Azure.Search.Documents/tests/Azure.Search.Documents.Tests.csproj:48]: Oh, sorry, this is removing the "FieldBuilder as a Sample" code, not FieldBuilder itself.  I was thinking that would get deleted but it's probably best to leave like you have for now.

File: sdk/search/Azure.Search.Documents/samples/Sample04_FieldBuilder.md
Comment (heaths) [reply sdk/search/Azure.Search.Documents/samples/Sample04_FieldBuilder.md:8]: The more I think about this, the more I think changing the title (and link description in the README) to set the expectation this is a workaround is probably better. That's really what was intended, but I agree that just calling the sample "FieldBuilder" is misleading. We already have a sample in the package README, so I don't think another is necessary here.

@AlexGhiondea, thoughts?

Comment (heaths) [reply sdk/search/Azure.Search.Documents/samples/Sample04_FieldBuilder.md]: Refocused the sample (file names, link descriptions, title, etc.) on `[FieldBuilderIgnore]`.

**Pull request merged**
